File: static/src/utils/roles.ts
import { User } from "../protectedRoute/authContext";

export const AppRole = {
    ADMIN: 'ADMIN',
    EDITOR: 'EDITOR',
    VIEWER: 'VIEWER'
};
export type SingleRoleType = keyof typeof AppRole
// Roles in order of privileges (highest first)
const rolesInOrder = [
    AppRole.ADMIN,
    AppRole.EDITOR,
    AppRole.VIEWER
];

export function getRolesChain(leastRole: SingleRoleType):SingleRoleType[] {
    const finalRoles = [];
    const leastRoleIndex = rolesInOrder.indexOf(leastRole);

    // Add roles up to (but not including) the least role
    for (let i = 0; i < leastRoleIndex; i++) {
        finalRoles.push(rolesInOrder[i]);
    }

    // Add the least role itself
    finalRoles.push(leastRole);

    return finalRoles as SingleRoleType[];
}
// TODO: enable permissions later
export function hasAccess(user: User, roleToCkeck:SingleRoleType) {
    const rolesChain = getRolesChain(roleToCkeck)
    return rolesChain.some((v) => user.role == v)
}

File: server/authentication/roles.py
from models import Role

# the first role in the array is having the heighest privellages
roles_in_order = [
    Role.ADMIN,
    Role.EDITOR,
    Role.VIEWER
]

def get_roles_chain(least_role):
    final_roles = []
    least_role_index = roles_in_order.index(least_role)
    for role in roles_in_order[0:least_role_index]:
        final_roles.append(role.name)
    final_roles.append(least_role.name)
    return final_roles
